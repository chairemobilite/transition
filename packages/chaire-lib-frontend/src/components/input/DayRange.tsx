/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React, { JSX } from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';
import _isEqual from 'lodash/isEqual';

import InputSelect from './InputSelect';

export type DayRangeProps = WithTranslation & {
    id: string;
    /** selection contains the number of the day, where 0 is monday */
    onChange: (selection: number[]) => void;
    days: number[];
    disabled?: boolean;
    showPeriodDropdown?: boolean;
};

type dayRangeType = 'custom' | 'week' | 'week-end' | 'all';

const periodsGroupValues = {
    week: [0, 1, 2, 3, 4],
    'week-end': [5, 6],
    all: [0, 1, 2, 3, 4, 5, 6]
};

const weekDays = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];

export class DayRange extends React.Component<DayRangeProps> {
    static defaultProps: Partial<DayRangeProps> = {
        disabled: false,
        showPeriodDropdown: true
    };

    constructor(props: DayRangeProps) {
        super(props);

        this.handleDayChange = this.handleDayChange.bind(this);
        this.handlePeriodChange = this.handlePeriodChange.bind(this);
    }

    handleDayChange(e, day: string) {
        const selection = [...this.props.days];
        const value = weekDays.indexOf(day);
        if (e.target.checked) {
            selection.push(value);
        } else {
            const idx = selection.indexOf(value);
            selection.splice(idx, 1);
        }
        this.props.onChange(selection);
    }

    handlePeriodChange(e) {
        const period = e.target.value;

        if (period !== 'custom') {
            const selection = periodsGroupValues[period];
            this.props.onChange(selection);
        }
    }

    renderButton(day: string, isChecked: boolean): JSX.Element {
        return (
            <div className="dayPickerOption" key={`${this.props.id}_${day}`}>
                <input
                    type="checkbox"
                    id={`${this.props.id}_${day}`}
                    value={day}
                    title={day}
                    name={day}
                    checked={isChecked}
                    onChange={(e) => this.handleDayChange(e, day)}
                    disabled={this.props.disabled}
                ></input>
                <label htmlFor={`${this.props.id}_${day}`}>
                    {this.props.t('main:dateTime:weekdaysAbbr:' + day.toLowerCase())}
                </label>
            </div>
        );
    }

    render() {
        const renderWeekDays: JSX.Element[] = [];
        const days = this.props.days;
        weekDays.forEach((day, index) => {
            const idx = days.indexOf(index);
            renderWeekDays.push(this.renderButton(day, idx >= 0));
        });

        let period = 'custom';
        for (const periodGroup in periodsGroupValues) {
            if (_isEqual(periodsGroupValues[periodGroup], days)) {
                period = periodGroup;
            }
        }

        const periodsGroupChoices = [
            { value: 'custom', label: this.props.t('main:dateTime:dayPeriod:Custom') },
            { value: 'week', label: this.props.t('main:dateTime:dayPeriod:Week') },
            { value: 'week-end', label: this.props.t('main:dateTime:dayPeriod:WeekEnd') },
            { value: 'all', label: this.props.t('main:dateTime:dayPeriod:All') }
        ];

        return (
            <div>
                {this.props.showPeriodDropdown && (
                    <div className="apptr__form-input-container _two-columns">
                        <label> {this.props.t('main:dateTime:Day')}</label>
                        <InputSelect
                            id={`${this.props.id}_periodGroup`}
                            t={this.props.t}
                            value={period}
                            choices={periodsGroupChoices}
                            noBlank={true}
                            disabled={this.props.disabled}
                            onValueChange={this.handlePeriodChange}
                        />
                    </div>
                )}
                <div className="tr__form-dayPicker">{renderWeekDays}</div>
            </div>
        );
    }
}

export default withTranslation('main')(DayRange);
