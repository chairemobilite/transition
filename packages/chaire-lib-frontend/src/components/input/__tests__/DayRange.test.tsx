/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import each from 'jest-each';

import DayRange from '../DayRange';

const mockOnChange = jest.fn();
const testId = 'DayRangeWidgetId';

test('All props', () => {
    const { container } = render(<DayRange
        id = {testId}
        onChange = {mockOnChange}
        days = {[0,1,2]}
        disabled = {false}
        showPeriodDropdown = {true}
    />);
    expect(container.firstChild).toMatchSnapshot();
});

test('Disabled', () => {
    const { container } = render(<DayRange
        id = {testId}
        onChange = {mockOnChange}
        days = {[0,1,2]}
        disabled = {true}
    />);
    expect(container.firstChild).toMatchSnapshot();
});

test('Do not show dropdown', () => {
    const { container } = render(<DayRange
        id = {testId}
        onChange = {mockOnChange}
        days = {[0,1,2]}
        showPeriodDropdown = {false}
    />);
    expect(container.firstChild).toMatchSnapshot();
});

describe('Period group selection change', () => {
    each([
        ['week', [0, 1, 2, 3, 4]],
        ['week-end', [5, 6]],
        ['all', [0, 1, 2, 3, 4, 5, 6]]
    ]).test('Period group selection for "%s"', (period, indices) => {
        const { container } = render(<DayRange
            id = {testId}
            onChange = {mockOnChange}
            days = {[0,1,2]}
            disabled = {false}
        />);

        mockOnChange.mockClear();

        // Validate initial values
        const periodGroupElement = container.querySelector(`#${testId}_periodGroup`) as HTMLInputElement;
        expect(periodGroupElement.value).toBe('custom');

        // Check the period and validate the choices
        fireEvent.change(periodGroupElement, { target: { value: period } });
        expect(mockOnChange).toHaveBeenCalledTimes(1);
        expect(mockOnChange).toHaveBeenCalledWith(indices);

    });

});

describe('Days selection change', () => {
    each([
        ['Monday', 0],
        ['Tuesday', 1],
        ['Wednesday', 2],
        ['Thursday', 3],
        ['Friday', 4],
        ['Saturday', 5],
        ['Sunday', 6],
    ]).test('Days selection for "%s"', async (day, index) => {

        const user = userEvent.setup();

        const { rerender }Â = render(<DayRange
            id = {testId}
            onChange = {mockOnChange}
            days = {[]}
            disabled = {false}
        />);

        mockOnChange.mockClear();
        expect(screen.getByTitle(day)).toBeInTheDocument();
        expect(screen.getByTitle(day)).not.toBeChecked();
        await user.click(screen.getByTitle(day) as HTMLInputElement);

        // FIXME: We used to check the checked state of the day in the dom, when the tests were using enzyme, but it was not working with testing-library

        // Validate that the onChanged has been called with the correct value
        expect(mockOnChange).toHaveBeenCalledTimes(1);
        expect(mockOnChange).toHaveBeenCalledWith([index]);

    });

});
