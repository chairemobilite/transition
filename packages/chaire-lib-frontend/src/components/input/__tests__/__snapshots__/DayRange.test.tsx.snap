// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All props 1`] = `
<div>
  <div
    class="apptr__form-input-container _two-columns"
  >
    <label>
       
      main:dateTime:Day
    </label>
    <select
      autocomplete="none"
      class="tr__form-input-select apptr__input _input-select"
      id="DayRangeWidgetId_periodGroup"
      name="DayRangeWidgetId_periodGroup"
      style="overflow: hidden; white-space: nowrap; text-overflow: ellipsis;"
    >
      <option
        value="custom"
      >
        main:dateTime:dayPeriod:Custom
      </option>
      <option
        value="week"
      >
        main:dateTime:dayPeriod:Week
      </option>
      <option
        value="week-end"
      >
        main:dateTime:dayPeriod:WeekEnd
      </option>
      <option
        value="all"
      >
        main:dateTime:dayPeriod:All
      </option>
    </select>
  </div>
  <div
    class="tr__form-dayPicker"
  >
    <div
      class="dayPickerOption"
    >
      <input
        checked=""
        id="DayRangeWidgetId_Monday"
        name="Monday"
        title="Monday"
        type="checkbox"
        value="Monday"
      />
      <label
        for="DayRangeWidgetId_Monday"
      >
        main:dateTime:weekdaysAbbr:monday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        checked=""
        id="DayRangeWidgetId_Tuesday"
        name="Tuesday"
        title="Tuesday"
        type="checkbox"
        value="Tuesday"
      />
      <label
        for="DayRangeWidgetId_Tuesday"
      >
        main:dateTime:weekdaysAbbr:tuesday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        checked=""
        id="DayRangeWidgetId_Wednesday"
        name="Wednesday"
        title="Wednesday"
        type="checkbox"
        value="Wednesday"
      />
      <label
        for="DayRangeWidgetId_Wednesday"
      >
        main:dateTime:weekdaysAbbr:wednesday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        id="DayRangeWidgetId_Thursday"
        name="Thursday"
        title="Thursday"
        type="checkbox"
        value="Thursday"
      />
      <label
        for="DayRangeWidgetId_Thursday"
      >
        main:dateTime:weekdaysAbbr:thursday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        id="DayRangeWidgetId_Friday"
        name="Friday"
        title="Friday"
        type="checkbox"
        value="Friday"
      />
      <label
        for="DayRangeWidgetId_Friday"
      >
        main:dateTime:weekdaysAbbr:friday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        id="DayRangeWidgetId_Saturday"
        name="Saturday"
        title="Saturday"
        type="checkbox"
        value="Saturday"
      />
      <label
        for="DayRangeWidgetId_Saturday"
      >
        main:dateTime:weekdaysAbbr:saturday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        id="DayRangeWidgetId_Sunday"
        name="Sunday"
        title="Sunday"
        type="checkbox"
        value="Sunday"
      />
      <label
        for="DayRangeWidgetId_Sunday"
      >
        main:dateTime:weekdaysAbbr:sunday
      </label>
    </div>
  </div>
</div>
`;

exports[`Disabled 1`] = `
<div>
  <div
    class="apptr__form-input-container _two-columns"
  >
    <label>
       
      main:dateTime:Day
    </label>
    <select
      autocomplete="none"
      class="tr__form-input-select apptr__input _input-select"
      disabled=""
      id="DayRangeWidgetId_periodGroup"
      name="DayRangeWidgetId_periodGroup"
      style="overflow: hidden; white-space: nowrap; text-overflow: ellipsis;"
    >
      <option
        value="custom"
      >
        main:dateTime:dayPeriod:Custom
      </option>
      <option
        value="week"
      >
        main:dateTime:dayPeriod:Week
      </option>
      <option
        value="week-end"
      >
        main:dateTime:dayPeriod:WeekEnd
      </option>
      <option
        value="all"
      >
        main:dateTime:dayPeriod:All
      </option>
    </select>
  </div>
  <div
    class="tr__form-dayPicker"
  >
    <div
      class="dayPickerOption"
    >
      <input
        checked=""
        disabled=""
        id="DayRangeWidgetId_Monday"
        name="Monday"
        title="Monday"
        type="checkbox"
        value="Monday"
      />
      <label
        for="DayRangeWidgetId_Monday"
      >
        main:dateTime:weekdaysAbbr:monday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        checked=""
        disabled=""
        id="DayRangeWidgetId_Tuesday"
        name="Tuesday"
        title="Tuesday"
        type="checkbox"
        value="Tuesday"
      />
      <label
        for="DayRangeWidgetId_Tuesday"
      >
        main:dateTime:weekdaysAbbr:tuesday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        checked=""
        disabled=""
        id="DayRangeWidgetId_Wednesday"
        name="Wednesday"
        title="Wednesday"
        type="checkbox"
        value="Wednesday"
      />
      <label
        for="DayRangeWidgetId_Wednesday"
      >
        main:dateTime:weekdaysAbbr:wednesday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        disabled=""
        id="DayRangeWidgetId_Thursday"
        name="Thursday"
        title="Thursday"
        type="checkbox"
        value="Thursday"
      />
      <label
        for="DayRangeWidgetId_Thursday"
      >
        main:dateTime:weekdaysAbbr:thursday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        disabled=""
        id="DayRangeWidgetId_Friday"
        name="Friday"
        title="Friday"
        type="checkbox"
        value="Friday"
      />
      <label
        for="DayRangeWidgetId_Friday"
      >
        main:dateTime:weekdaysAbbr:friday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        disabled=""
        id="DayRangeWidgetId_Saturday"
        name="Saturday"
        title="Saturday"
        type="checkbox"
        value="Saturday"
      />
      <label
        for="DayRangeWidgetId_Saturday"
      >
        main:dateTime:weekdaysAbbr:saturday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        disabled=""
        id="DayRangeWidgetId_Sunday"
        name="Sunday"
        title="Sunday"
        type="checkbox"
        value="Sunday"
      />
      <label
        for="DayRangeWidgetId_Sunday"
      >
        main:dateTime:weekdaysAbbr:sunday
      </label>
    </div>
  </div>
</div>
`;

exports[`Do not show dropdown 1`] = `
<div>
  <div
    class="tr__form-dayPicker"
  >
    <div
      class="dayPickerOption"
    >
      <input
        checked=""
        id="DayRangeWidgetId_Monday"
        name="Monday"
        title="Monday"
        type="checkbox"
        value="Monday"
      />
      <label
        for="DayRangeWidgetId_Monday"
      >
        main:dateTime:weekdaysAbbr:monday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        checked=""
        id="DayRangeWidgetId_Tuesday"
        name="Tuesday"
        title="Tuesday"
        type="checkbox"
        value="Tuesday"
      />
      <label
        for="DayRangeWidgetId_Tuesday"
      >
        main:dateTime:weekdaysAbbr:tuesday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        checked=""
        id="DayRangeWidgetId_Wednesday"
        name="Wednesday"
        title="Wednesday"
        type="checkbox"
        value="Wednesday"
      />
      <label
        for="DayRangeWidgetId_Wednesday"
      >
        main:dateTime:weekdaysAbbr:wednesday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        id="DayRangeWidgetId_Thursday"
        name="Thursday"
        title="Thursday"
        type="checkbox"
        value="Thursday"
      />
      <label
        for="DayRangeWidgetId_Thursday"
      >
        main:dateTime:weekdaysAbbr:thursday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        id="DayRangeWidgetId_Friday"
        name="Friday"
        title="Friday"
        type="checkbox"
        value="Friday"
      />
      <label
        for="DayRangeWidgetId_Friday"
      >
        main:dateTime:weekdaysAbbr:friday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        id="DayRangeWidgetId_Saturday"
        name="Saturday"
        title="Saturday"
        type="checkbox"
        value="Saturday"
      />
      <label
        for="DayRangeWidgetId_Saturday"
      >
        main:dateTime:weekdaysAbbr:saturday
      </label>
    </div>
    <div
      class="dayPickerOption"
    >
      <input
        id="DayRangeWidgetId_Sunday"
        name="Sunday"
        title="Sunday"
        type="checkbox"
        value="Sunday"
      />
      <label
        for="DayRangeWidgetId_Sunday"
      >
        main:dateTime:weekdaysAbbr:sunday
      </label>
    </div>
  </div>
</div>
`;
